datafusion/core/src/physical_plan/row_converter.rs

```diff
@@ -0,0 +1,292 @@
+const LOW_CARDINALITY_THRESHOLD: usize = 10;
+
+/// Wrapper around an [`arrow::row`] [`RowConverter`] that disables
+/// dictionary preservation for high cardinality columns, based on the
+/// observed cardinalities in the first columns converted.
+///
+/// ## Background
+///
+/// By default, the [`RowConverter`] interns (and keeps a copy of) all
+/// values from [`DictionaryArray`]s. For low cardinality columns
+/// (with few distinct values) this approach is both faster and more
+/// memory efficient. However for high cardinality coumns it is slower
+/// and requires more memory. In certain degenerate cases, such as
+/// columns of nearly unique values, the `RowConverter` will keep a
+/// copy of the entire column.
+///
+/// See <https://github.com/apache/arrow-datafusion/issues/7200> for
+/// more details
+#[derive(Debug)]
+pub enum CardinalityAwareRowConverter {
+    /// Converter is newly initialized, and hasn't yet seen data
+    New {
+        /// Defines the Row conversion
+        fields: Vec<SortField>,
+    },
+    /// Converter has seen data and can convert [`Array`]s to/from
+    /// [`Rows`]
+    Converting {
+        /// Underlying converter
+        converter: RowConverter,
+        /// if preserve_dictionaries is disabled, the output type can be
+        /// different than input.
+        output_types: Vec<DataType>,
+    },
+}
+
+impl CardinalityAwareRowConverter {
+    pub fn new(fields: Vec<SortField>) -> Result<Self> {
+        Ok(Self::New { fields })
+    }
+
+    /// Returns the memory size of the underlying [`RowConverter`] if
+    /// any.
+    pub fn size(&self) -> usize {
+        match self {
+            Self::New { .. } => 0,
+            Self::Converting { converter, .. } => converter.size(),
+        }
+    }
+
+    pub fn empty_rows(&self, row_capacity: usize, data_capacity: usize) -> Result<Rows> {
+        match self {
+            Self::New { .. } => internal_err!(
+                "CardinalityAwareRowConverter has not converted any rows yet"
+            ),
+            Self::Converting { converter, .. } => {
+                Ok(converter.empty_rows(row_capacity, data_capacity))
+            }
+        }
+    }
+
+    pub fn convert_rows<'a, I>(&self, rows: I) -> Result<Vec<ArrayRef>>
+    where
+        I: IntoIterator<Item = Row<'a>>,
+    {
+        match self {
+            Self::New { .. } => internal_err!(
+                "CardinalityAwareRowConverter has not converted any rows yet"
+            ),
+            Self::Converting {
+                converter,
+                output_types,
+            } => {
+                // Cast output type if needed
+                let output = converter
+                    .convert_rows(rows)?
+                    .into_iter()
+                    .zip(output_types.iter())
+                    .map(|(arr, output_type)| {
+                        if arr.data_type() != output_type {
+                            Ok(arrow::compute::cast(&arr, output_type)?)
```

Comment:
It is perhaps worth highlighting that if the dictionary is indeed high cardinality, this conversion will be quite expensive. I don't have a good solution for this

Reply:
One potential solution is to move it upstream in arrow-rs and avoid the intermediate copy, perhaps. I have filed a ticket https://github.com/apache/arrow-rs/issues/4813 to track this idea